### PR TITLE
feat(swap): add Swap v2 integration

### PR DESCRIPTION
**Pull request by onyb**
Resolves https://github.com/brave/brave-browser/issues/21704
Resolves https://github.com/brave/brave-browser/issues/23161
Resolves https://github.com/brave/brave-browser/issues/21565
Resolves https://github.com/brave/brave-browser/issues/21761
Resolves https://github.com/brave/brave-browser/issues/26632

| Security review | https://github.com/brave/security/issues/1100 |
-|-

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Visit `brave://wallet/swap` to test out Swap V2.

### CODE REVIEW

File: third_party/npm_@brave_swap-interface/README.chromium [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Name: @brave/swap-interface
+URL: https://github.com/brave/swap
```
Comment (diracdeltas) [inline third_party/npm_@brave_swap-interface/README.chromium:2]: seems this repo needs some security stuff set up, ex: https://github.com/brave/swap/pull/100

i also recommend a github action that runs `npm audit`

cc @kdenhartog 

probably both the swap repo and this PR should have a security review (can be the same ticket)

**Comment by diracdeltas** [comment]
i see a socket alert here: https://github.com/brave/brave-core/pull/15762#issuecomment-1300129142

it doesn't look like a huge deal, but is there a newer version of the package that doesn't cause this alert?

also this repo should not have a dependency on a package in brave-experiments: https://github.com/brave/swap/blob/master/interface/package.json#L48

it needs to be moved to brave org before being used in a production project

**Comment by Douglashdaniel** [comment]
> 



> i see a socket alert here: [#15762 (comment)](https://github.com/brave/brave-core/pull/15762#issuecomment-1300129142)
> 
> it doesn't look like a huge deal, but is there a newer version of the package that doesn't cause this alert? UDPATE: looks like no
> 
> also this repo should not have a dependency on a package in brave-experiments: https://github.com/brave/swap/blob/master/interface/package.json#L48
> 
> it needs to be moved to brave org before being used in a production project

@petemill is this possible at the moment? 🙏

File: third_party/npm_@brave_swap-interface/README.chromium (hunk 1 above)
Comment (kdenhartog) [reply third_party/npm_@brave_swap-interface/README.chromium:2]: Yeah a security ticket to manage this would be good. Please also assign it to me @onyb

**Comment by petemill** [comment]
> > it needs to be moved to brave org before being used in a production project
> 
> @petemill is this possible at the moment? 🙏

Yeah I'm just about to start this process! 😄

File: third_party/npm_@brave_swap-interface/README.chromium (hunk 1 above)
Comment (onyb) [reply third_party/npm_@brave_swap-interface/README.chromium:2]: Done: https://github.com/brave/security/issues/1100 

@kdenhartog

**Review by kdenhartog — COMMENTED** [review]
static analysis looks good, gonna do some quick dynamic testing today and then should be G2G. Left a few nits I spotted in here is all

**Comment by kdenhartog** [comment]
Did some dynamic testing and it all seems to be have as expected for the most part. I did notice that the error message if the nonce value is intentionally set incorrectly produces a bad message. 

https://user-images.githubusercontent.com/23125059/202654300-c3fceae1-0b08-44d8-aaa7-2684e7ecd7a6.mov

I did also check that the socket security alert isn't appearing when using locale that [should trigger it](https://github.com/styled-components/styled-components/blob/legacy-v5/packages/styled-components/postinstall.js#L78) when on the swap screen appears so I think we're fine to use the dependency. 

Last thing left should be finishing moving the Leo package over so that we don't have brave-experiments dependencies and I think this one will be G2G then.

File: package.json [hunk 2]
```diff
@@ -294,6 +294,7 @@
     "acorn": "8.0.5",
     "babel-loader": "8.2.3",
     "brave-ui": "github:brave/brave-ui#e4c049880199ac652b8f69610b8b70d1cd6d906f",
+    "@brave/swap-interface": "0.11.7",
```
Comment (kdenhartog) [inline package.json:297]:
```suggestion
    "@brave/swap-interface": "0.12.1",
```

**Review by kdenhartog — APPROVED** [review]
Pending version update here that uses brave org leo package this one is G2G from a security perspective

**Comment by onyb** [comment]
@kdenhartog I think the issue with the incorrect error message is known - if I'm not mistaken @srirambv had encountered this recently.

Also going to ignore the Socket Security warning.

File: components/brave_wallet/browser/swap_response_parser.cc [hunk 3]
```diff
@@ -89,6 +104,16 @@ mojom::SwapResponsePtr ParseSwapResponse(const std::string& json,
       swap_response_value->sell_token_to_eth_rate;
   swap_response->buy_token_to_eth_rate =
       swap_response_value->buy_token_to_eth_rate;
+  swap_response->estimated_price_impact =
+      swap_response_value->estimated_price_impact;
+
+  for (const auto& source_value : swap_response_value->sources) {
+    mojom::ZeroExSource source;
```
Comment (supermassive) [inline components/brave_wallet/browser/swap_response_parser.cc:111]: nit: mojom::ZeroExSource::New() creates a ZeroExSourcePtr which could be moved into vector so no need in cloning.

Comment (supermassive) [reply components/brave_wallet/browser/swap_response_parser.cc:111]: Actually could be inlined as push_back(mojom::ZeroExSource::New(source_value.name, source_value.proportion))

File: components/brave_wallet/browser/swap_service_unittest.cc [hunk 4]
```diff
@@ -349,6 +369,11 @@ TEST_F(SwapServiceUnitTest, GetTransactionPayload) {
       "0xdef1c0ded9bec7f1a1670819833240f027b25eff";
   expected_swap_response->sell_token_to_eth_rate = "1900.44962824532464391";
   expected_swap_response->buy_token_to_eth_rate = "1";
+  expected_swap_response->estimated_price_impact = "0.7232";
+  auto source = brave_wallet::mojom::ZeroExSource::New();
+  source->name = "Uniswap_V2";
+  source->proportion = "1";
+  expected_swap_response->sources.push_back(source.Clone());
```
Comment (supermassive) [inline components/brave_wallet/browser/swap_service_unittest.cc:376]: inline or move

File: components/brave_wallet/browser/swap_response_parser.cc (hunk 3 above)
Comment (onyb) [reply components/brave_wallet/browser/swap_response_parser.cc]: Fixed https://github.com/brave/brave-core/pull/15762/commits/c4d08e374d30735c892cecaeeeb0f6b5c68bcad9.

**Review by Douglashdaniel — APPROVED** [review]
++

**Comment by supermassive** [comment]
components/brave_wallet/ LGTM

**Review by supermassive — APPROVED** [review]

**Review by diracdeltas — APPROVED** [review]
unblocking deps review and approving ignoring the protestware warning on socket

**Pull request merged**